xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,378 @@
+
+  @VisibleForTesting
+  final class ConfigSelector extends InternalConfigSelector {
+    @Override
+    public Result selectConfig(PickSubchannelArgs args) {
+      // Index ASCII headers by keys.
+      Map<String, Iterable<String>> asciiHeaders = new HashMap<>();
+      Metadata headers = args.getHeaders();
+      for (String headerName : headers.keys()) {
+        if (headerName.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+          continue;
+        }
+        Metadata.Key<String> key = Metadata.Key.of(headerName, Metadata.ASCII_STRING_MARSHALLER);
+        asciiHeaders.put(headerName, headers.getAll(key));
+      }
+      String cluster = null;
+      Route selectedRoute = null;
+      do {
+        for (Route route : routes) {
+          if (route.getRouteMatch().matches(
+              "/" + args.getMethodDescriptor().getFullMethodName(), asciiHeaders)) {
+            selectedRoute = route;
+            break;
+          }
+        }
+        if (selectedRoute == null) {
+          return Result.forError(Status.UNAVAILABLE.withDescription("Failed to route the RPC"));
+        }
+        RouteAction action = selectedRoute.getRouteAction();
+        if (action.getCluster() != null) {
+          cluster = action.getCluster();
+        } else if (action.getWeightedCluster() != null) {
+          int totalWeight = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            totalWeight += weightedCluster.getWeight();
+          }
+          int select = random.nextInt(totalWeight);
+          int accumulator = 0;
+          for (ClusterWeight weightedCluster : action.getWeightedCluster()) {
+            accumulator += weightedCluster.getWeight();
+            if (select < accumulator) {
+              cluster = weightedCluster.getName();
+              break;
+            }
+          }
+        }
+        if (cluster == null) {  // should not happen if routing rules are configured correctly
+          return Result.forError(
+              Status.UNAVAILABLE.withDescription("Failed to route the RPC with selected action"));
+        }
+      } while (!retainCluster(cluster));
+      final String finalCluster = cluster;
+      class SelectionCompleted implements Runnable {
+        @Override
+        public void run() {
+          releaseCluster(finalCluster);
+        }
+      }
+
+      Map<String, ?> rawServiceConfig =
+          generateServiceConfigWithMethodConfig(
+              args.getMethodDescriptor().getFullMethodName(),
+              selectedRoute.getRouteAction().getTimeoutNano());
+      if (logger.isLoggable(XdsLogLevel.INFO)) {
+        logger.log(XdsLogLevel.INFO,
+            "Generated service config (method config):\n{0}", new Gson().toJson(rawServiceConfig));
+      }
+      ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+      Object config = parsedServiceConfig.getConfig();
+      if (config == null) {
+        throw new AssertionError(
```

Comment:
Right. I had discussion with @dapengzhang0 at https://github.com/grpc/grpc-java/pull/7275#discussion_r465410835 about this. I was more leaning towards making the config in the result of `selectConfig()` a `ConfigOrError` (instead of just the config `Object`) such that the client call has the option to fail the RPC or proceed with no method config. In either case, the client call needs to invoke the callback after it.

The other option as you mentioned here is to fail the `selectConfig()` operation, even if the call can be routed to a cluster.

I do not have strong opinion between the two, while slightly leaning towards the first one as it seems to be more flexible for client call and the handling is slightly cleaner (i.e., it is always the call that releases the cluster).

Reply:
> I was more leaning towards making the config in the result of selectConfig() a ConfigOrError (instead of just the config Object) such that the client call has the option to fail the RPC or proceed with no method config.

Why would the client call have a better idea of how to proceed? It is very clear in this code that we don't know the configuration to use. I don't understand how client call could fabricate some configuration that would make things "okay."

> The other option as you mentioned here is to fail the selectConfig() operation, even if the call can be routed to a cluster.

It can't be routed to the cluster though. We may know which cluster, but we don't know what configuration to use to that cluster. So we can't send any traffic to it. For all we know there is fault injection configuration to fail 100% of the calls.

Reply:
> It is very clear in this code that we don't know the configuration to use. I don't understand how client call could fabricate some configuration that would make things "okay."

Basically, the `Result` contains the cluster name (which we may know as you said even if the configuration is not known), and a `ConfigOrError` that contains the error for fail to get the configuration. So the client call takes the responsibility of whether or not to fail the call for cases that cluster name is known while configuration is invalid. Doesn't this sound better?

Reply:
> So the client call takes the responsibility of whether or not to fail the call for cases that cluster name is known while configuration is invalid.

What is the advantage? I was claiming that _not failing_ the call is broken. We _must_ fail the call.

Reply:
Just some flexibility the API provides to client calls. If the answer is call should be failed for sure, then we do not need such flexibility. I will fix it.

Reply:
Fixed. It returns an error Result (and release the cluster in-place) if parsing config encounter error.